src/Microsoft.ML.FastTree/TreeEnsembleFeaturizer.cs

```diff
@@ -735,14 +735,14 @@ public static IDataTransform CreateForEntryPoint(IHostEnvironment env, Arguments
             bool identity;
             var converter = Conversions.Instance.GetStandardConversion<TInput, ulong>(type, dstType, out identity);
             var isNa = Conversions.Instance.GetIsNAPredicate<TInput>(type);
-            ulong temp = 0;
 
             ValueMapper<TInput, Single> mapper;
             if (seed == 0)
             {
                 mapper =
                     (in TInput src, ref Single dst) =>
                     {
+                        ulong temp = 0;
```

Comment:
Very nice find! :-) Can you please add some comments about this change? Otherwise some future developer might want to "optimize" this and move it back. 

Also, this pattern deserves more investigation. Do we have other mappers that are using variables set outside the mapper function definition?

Reply:
Yes, sure, I will add comments on this.

---
In reply to: [394130751](https://github.com/dotnet/machinelearning/pull/4950#discussion_r394130751) [](ancestors = 394130751)

Reply:
@sharwell do you have any tool recommendation to detect similar race condition issue either static code analyze or run time detection?

---
In reply to: [394131983](https://github.com/dotnet/machinelearning/pull/4950#discussion_r394131983) [](ancestors = 394131983,394130751)

Reply:
No similar issue found from other mappers but mapper are not only function used in multi-threading condition and it is impossible to view for all similar issue.

---
In reply to: [394130751](https://github.com/dotnet/machinelearning/pull/4950#discussion_r394130751) [](ancestors = 394130751)

Reply:
I'm not aware of such a tool. It would be easier to write a more targeted tool that identified variables captured by a lambda so they could be reviewed for cases like this.

Reply:
Got it, thanks

---
In reply to: [394334398](https://github.com/dotnet/machinelearning/pull/4950#discussion_r394334398) [](ancestors = 394334398)